### PR TITLE
[specs][kimchi] spec the verifier

### PR DESCRIPTION
**Pull request by mimoo**
specifying the verifier.

Also made a few changes as I was doing that:

* removed an assert, and made the verifier return its own error type
* created a `verify` for a single proof, applied it to most tests
* added some doc / renamed some variables

### CODE REVIEW

File: book/src/specs/kimchi.md [hunk 1]
```diff
@@ -838,6 +838,71 @@ The prover then follows the following steps to create the proof:
...
 ### Proof Verification
 
+We define two helper algorithms below, used in the batch verification of proofs. 
+
+
+#### Fiat-Shamir argument
+
+We run the following algorithm:
+
+10. Sample $\alpha'$ with the Fq-Sponge.
```
Comment (jspada) [inline book/src/specs/kimchi.md:848]: Starting on step `10`?

File: book/src/specs/kimchi.md [hunk 2]
```diff
@@ -838,6 +838,71 @@ The prover then follows the following steps to create the proof:
...
+11. Derive $\alpha$ from $\alpha'$ using the endomorphism (TODO: details).
+12. Enforce that the length of the $t$ commitment is of size `PERMUTS`.
+13. absorb the commitment to the quotient polynomial $t$ into the argument and sample zeta.
```
Comment (jspada) [inline book/src/specs/kimchi.md:851]: Capital for consistency

$\zeta$

File: book/src/specs/kimchi.md [hunk 3]
```diff
@@ -838,6 +838,71 @@ The prover then follows the following steps to create the proof:
...
+21. Sample $v'$ with the Fr-Sponge.
+22. Derive $v$ from $v'$ using the endomorphism (TODO: specify).
+23. Sample $u'$ with the Fr-Sponge.
+24. Derive $u$ from $u'$ using the endomorphism (TODO: specify).
+25. Create a list of all polynomials that have an evaluation proof.
+26. Compute the evaluation of $ft(\zeta)$.
+
+#### Partial verification
+
+For every proof we want to verify, we deffer the proof opening to the very end.
```
Comment (jspada) [inline book/src/specs/kimchi.md:875]: defer

File: book/src/specs/kimchi.md [hunk 4]
```diff
@@ -838,6 +838,71 @@ The prover then follows the following steps to create the proof:
...
+This allows us to potentially batch verify a number of partially verified proofs.
+Essentially, this steps verify that $f(\zeta) = t(\zeta) * Z_H(\zeta)$.
```
Comment (jspada) [inline book/src/specs/kimchi.md:877]: verifies

File: book/src/specs/kimchi.md [hunk 5]
```diff
@@ -838,6 +838,71 @@ The prover then follows the following steps to create the proof:
...
+
+1. Commit to the negated public input polynomial.
+2. Run the [Fiat-Shamir argument](#fiat-shamir-argument).
+3. combine the chunked polynomials' evaluations
```
Comment (jspada) [inline book/src/specs/kimchi.md:881]: Capital

File: book/src/specs/kimchi.md [hunk 6]
```diff
@@ -838,6 +838,71 @@ The prover then follows the following steps to create the proof:
...
+   (TODO: most likely only the quotient polynomial is chunked)
+   with the right powers of $\zeta^n$ and $(\zeta * \omega)^n$.
+4. Compute the commitment to the linearized polynomial $f$.
+5. Compute the (chuncked) commitment of $ft$
+   (see [Maller's optimization](../crypto/plonk/maller_15.html)).
+6. List the polynomial commitments, and their associated evaluations,
+   That are associated to the aggregated evaluation proof in the proof:
```
Comment (jspada) [inline book/src/specs/kimchi.md:888]: lowercase that

Comment (jspada) [reply book/src/specs/kimchi.md:888]: So you're saying you "put a proof in your proof so you can prove while you prove?"

File: book/src/specs/kimchi.md [hunk 7]
```diff
@@ -838,6 +838,71 @@ The prover then follows the following steps to create the proof:
...
+    - recursion
+    - public input commitment
+    - ft commitment (chunks of it)
+    - permutation commitment
+    - index commitments that use the coefficients
+    - witness commitments
+    - sigma commitments
+#### Batch verification of proofs
+
+Below, we define the steps to follow a number of proofs
```
Comment (jspada) [inline book/src/specs/kimchi.md:898]: for a number of proofs?

**Review by jspada — APPROVED** [review]
Approved with some typos to fix.

File: book/src/specs/kimchi.md (hunk 1 above)
Comment (mimoo) [reply book/src/specs/kimchi.md:848]: good catch! was missing some doc comments

File: book/src/specs/kimchi.md (hunk 6 above)
Comment (mimoo) [reply book/src/specs/kimchi.md:888]: there's definitely a proof in the proof

File: kimchi/src/tests/generic.rs [hunk 8]
```diff
@@ -73,19 +73,15 @@ fn verify_proof(gates: Vec<CircuitGate<Fp>>, witness: [Vec<Fp>; COLUMNS], public
...
+        &index,
+        vec![prev],
+    )
+    .unwrap();
 
     // verify the proof
     let verifier_index = index.verifier_index();
-    let batch: Vec<_> = batch.iter().map(|proof| (&verifier_index, proof)).collect();
-    batch_verify::<Affine, BaseSponge, ScalarSponge>(&group_map, &batch).unwrap();
+    verify::<Affine, BaseSponge, ScalarSponge>(&group_map, &verifier_index, &proof).unwrap();
```
[the remark this answers is not in the document]
Comment (mimoo) [reply kimchi/src/tests/generic.rs:86]: there's a single proof, so let's just use verify instead of batch_verify which is for several proofs

**Pull request merged**